### PR TITLE
Improve MAV_CMD parameter information

### PR DESCRIPTION
**Pull request by amilcarlucas**
This is a followup of #7 and goes together with mavlink/mavlink#623

### CODE REVIEW

**Comment by amilcarlucas** [comment]
CI tests pass, this does not change the generated code, nor does it change the generated documentation.
I would really like to get feedback on this one. @DonLakeFlyer @LorenzMeier @dagar

**Comment by dagar** [comment]
Overall looks good to me. From here I'd want to go through a few examples of updating common to see how well it works in practice.

Could you elaborate on the intended use of hasLocation and isDestination? What about other coordinate frames? I find the use of location in MAV_CMDs pretty confusing. With COMMAND_INT there's a field for the MAV_FRAME (https://mavlink.io/en/messages/common.html#COMMAND_INT), but I don't know how widespread the usage is.

**Comment by DonLakeFlyer** [comment]
> Could you elaborate on the intended use of hasLocation and isDestination?

hasLocation - Command specifies a lat/lon/alt
isDestination - Command specified a lat/lon/alt and the vehicle flies through the point

@amilcarlucas QGC also has the concept of specifiesAltitudeOnly for a command. In this case only param 7 has altitude, but 6/7 are NOT lat/lon.

**Comment by dagar** [comment]
What about different frames? There are already some that could be either global or local, some that are local only, relative alt, etc.

**Comment by DonLakeFlyer** [comment]
> What about different frames? There are already some that could be either global or local, some that are local only, relative alt, etc.

The frame used when COMMAND_LONG is used should be in there somewhere.

**Comment by peterbarker** [comment]
On Tue, 14 Aug 2018, Don Gagne wrote:

>
>       Could you elaborate on the intended use of hasLocation and isDestination?
> 
> hasLocation - Command specifies a lat/lon/alt
> isDestination - Command specified a lat/lon/alt and the vehicle flies through the point

What about the "kinda specifies location" commands like NAV_CMD_NAV_LAND - 
where ArduCopter will wander over to a location if it is specified before 
landing, but if one isn't specified it just lands?

**Comment by DonLakeFlyer** [comment]
> What about the "kinda specifies location" commands like NAV_CMD_NAV_LAND -
> where ArduCopter will wander over to a location if it is specified before
> landing, but if one isn't specified it just lands?

Is that the lat/lon = 0 special case stuff. From a QGC standpoint I have no idea how to handle that sort of thing with a decent ui.

**Comment by peterbarker** [comment]
On Tue, 14 Aug 2018, Don Gagne wrote:

> Is that the lat/lon = 0 special case stuff. From a QGC standpoint I have no idea how to handle that sort of thing with a decent ui.

Yes it is.  I'm not saying it's a good thing, but it is a thing :-)

I guess the markup could be "optional_location" or something?

File: generator/mavschema.xsd [hunk 1]
```diff
@@ -9,15 +9,14 @@
...
 <!-- definition of attributes -->
-<xs:attribute name="name"> <!-- enum,entry,message,field -->
+<xs:attribute name="name"> <!-- used in enum,entry,message,field elements -->
 </xs:attribute>
-<xs:attribute name="index" type="xs:unsignedByte"/> <!-- param -->
-<xs:attribute name="id" type="xs:unsignedShort"/> <!-- message -->
-<xs:attribute name="print_format" type="xs:string"/> <!-- field -->
-<xs:attribute name="enum" type="xs:string"/>
-<xs:attribute name="display" type="xs:string"/>
+<xs:attribute name="id" type="xs:unsignedShort"/> <!-- used in message elements -->
```
Comment (hamishwillee) [inline generator/mavschema.xsd]: Can't a message id be up to 24 bits? This is 16 bits.

**Comment by DonLakeFlyer** [comment]
> Should the "vehicle flies through the point" really be part of this? ie what about case where vehicle doesn't fly through point.

The point of this information is so that a ground station can use it to draw waypoint lines on a map. For that it needs to know whether an item specifies a location, and if the vehicle does or does not fly through that location.

**Comment by DonLakeFlyer** [comment]
> I guess the markup could be "optional_location" or something?

I guess. But making that part of the standard mavlink spec makes my skin crawl a bit. I wouldn't want new examples of this to show up if at all possible.

**Comment by amilcarlucas** [comment]
@hamishwillee most of the discussion on the attributes of this PR are on mavlink/mavlink#623

File: generator/mavschema.xsd (hunk 1 above)
Comment (hamishwillee) [reply generator/mavschema.xsd]: @amilcarlucas ?

Comment (amilcarlucas) [reply generator/mavschema.xsd]: Very good point. As you can see that has been a "xs:unsignedShort" even before my changes.

So ... That is something that should be fixed by a future PR.

Comment (hamishwillee) [reply generator/mavschema.xsd]: Fair enough! Let's try remember to post one when this has gone in :-).

Comment (hamishwillee) [reply generator/mavschema.xsd]: Created https://github.com/ArduPilot/pymavlink/issues/216 for tracking

**Comment by amilcarlucas** [comment]
@tridge  can I merge this ?

File: generator/mavschema.xsd [hunk 2]
```diff
@@ -135,10 +134,29 @@
   </xs:restriction>
 </xs:simpleType>
 
+<!-- definition param elements attributes (like the ones used on MAV_CMD for example) -->
+<xs:attribute name="index" type="xs:unsignedByte"/>                    <!-- parameter index -->
+<xs:attribute name="label" type="xs:string"/>                          <!-- parameter label (name) -->
+<xs:attribute name="decimalPlaces" type="xs:unsignedByte"/>            <!-- parameter number of decimal places to be displayed in the GUI -->
```
Comment (amilcarlucas) [inline generator/mavschema.xsd]: I guess this should be replaced with `print_format`. That way it reuses an existing concept. There is no need to introduce a new one IMHO.

File: generator/mavschema.xsd [hunk 3]
```diff
@@ -135,10 +134,29 @@
...
+<xs:attribute name="isDestination" type="xs:boolean" default="true"/>  <!-- entry is a destination -->
+
 <!-- definition of complex elements -->
 <xs:element name="param">
     <xs:complexType mixed="true">
         <xs:attribute ref="index" use="required"/>
+        <xs:attribute ref="label"/>
+        <xs:attribute ref="units"/>
+        <xs:attribute ref="enum" />
+        <xs:attribute ref="decimalPlaces"/>
```
Comment (amilcarlucas) [inline generator/mavschema.xsd]: I guess this should be replaced with `print_format`. That way it reuses an existing concept. There is no need to introduce a new one IMHO.

**Comment by DonLakeFlyer** [comment]
In QGC the concept of decimal places is not just used to display values to the user but there are various places in the code where the actual precision of the value is needed to be known. For those the number of decimal places is required. This usage is currently only for parameter values not MAV_CMD, but I could see this coming up for that as well.

The specific case is for these slider controls:
screen shot 2018-09-29 at 10 07 37 am

In order to build a set of value before/after the current value the precision of the value being modified needs to be known such that those values match the precision. I could see the same thing coming up for a UI which is used to build up COMMAND_LONG calls.

**Comment by hamishwillee** [comment]
@amilcarlucas Can you chase @tridge about this one in the ArduPilot dev call or recommend someone else with the authority and background to approve this?

**Comment by amilcarlucas** [comment]
@DonLakeFlyer well, print_format can be used to achieve the same thing.
@hamishwillee yes, I will do that.

**Comment by DonLakeFlyer** [comment]
What's the definition of print_format and how would it support the same thing?

**Comment by DonLakeFlyer** [comment]
Having to write a generic print format parser is not something anyone would really want to have to implement.

**Comment by amilcarlucas** [comment]
The point is that the mavlink spec already uses `print_format` since a couple of years.
So you (and/or other GCSs) probably have that implemented already anyways.

**Pull request merged**